File: tests/rosetta/transpiler/php/ackermann-function-3.php
<?php
function pow_big($base, $exp) {
  global $bit_len, $err, $ackermann2, $show, $main;
  $result = 1;
  $b = $base;
  $e = $exp;
  while ($e > 0) {
  if ($e % 2 == 1) {
  $result = $result * $b;
}
  $b = $b * $b;
  $e = intval((intdiv($e, 2)));
};
  return $result;
}
function bit_len($x) {
  global $pow_big, $err, $ackermann2, $show, $main;
  $n = $x;
  $c = 0;
  while ($n > 0) {
  $n = $n / 2;
  $c = $c + 1;
};
  return $c;
}
$err = "";
function ackermann2($m, $n) {
  global $pow_big, $bit_len, $err, $show, $main;
  if ($err != "") {
  return 0;
}
  if ($m <= 3) {
  $mi = intval($m);
  if ($mi == 0) {
  return $n + 1;
};
  if ($mi == 1) {
  return $n + 2;
};
  if ($mi == 2) {
  return 2 * $n + 3;
};
  if ($mi == 3) {
  $nb = bit_len($n);
  if ($nb > 64) {
  $err = "A(m,n) had n of " . json_encode($nb, 1344) . " bits; too large";
  return 0;
};
  $r = pow_big(2, intval($n));
  return 8 * $r - 3;
};
}
  if (bit_len($n) == 0) {
  return ackermann2($m - (1), 1);
}
  return ackermann2($m - (1), ackermann2($m, $n - (1)));
}
function show($m, $n) {
  global $pow_big, $bit_len, $err, $ackermann2, $main;
  $err = "";
  $res = ackermann2($m, $n);
  if ($err != "") {
  echo "A(" . json_encode($m, 1344) . ", " . json_encode($n, 1344) . ") = Error: " . $err, PHP_EOL;
  return;
}
  if (bit_len($res) <= 256) {
  echo "A(" . json_encode($m, 1344) . ", " . json_encode($n, 1344) . ") = " . json_encode($res, 1344), PHP_EOL;
} else {
  $s = json_encode($res, 1344);
  $pre = substr($s, 0, 20 - 0);
  $suf = substr($s, strlen($s) - 20, strlen($s) - strlen($s) - 20);
  echo "A(" . json_encode($m, 1344) . ", " . json_encode($n, 1344) . ") = " . json_encode(strlen($s), 1344) . " digits starting/ending with: " . $pre . "..." . $suf, PHP_EOL;
}
}
function main() {
  global $pow_big, $bit_len, $err, $ackermann2, $show;
  show(0, 0);
  show(1, 2);
  show(2, 4);
  show(3, 100);
  show(3, 1000000);
  show(4, 1);
  show(4, 2);
  show(4, 3);
}
main();

File: tests/rosetta/transpiler/FS/ackermann-function-3.fs
// Generated 2025-07-25 13:04 +0700

exception Return

let rec pow_big (``base``: bigint) (exp: int) =
    let mutable __ret : bigint = Unchecked.defaultof<bigint>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable result: bigint = 1 :?> bigint
        let mutable b: bigint = ``base``
        let mutable e: int = exp
        while e > 0 do
            if (e % 2) = 1 then
                result <- result * b
            b <- b * b
            e <- int (e / 2)
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and bit_len (x: bigint) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable x = x
    try
        let mutable n: bigint = x
        let mutable c: int = 0
        while n > 0 do
            n <- n / 2
            c <- c + 1
        __ret <- c
        raise Return
        __ret
    with
        | Return -> __ret
let mutable err: string = ""
let rec ackermann2 (m: bigint) (n: bigint) =
    let mutable __ret : bigint = Unchecked.defaultof<bigint>
    let mutable m = m
    let mutable n = n
    try
        if err <> "" then
            __ret <- 0 :?> bigint
            raise Return
        if m <= 3 then
            let mi = int m
            if mi = 0 then
                __ret <- n + 1
                raise Return
            if mi = 1 then
                __ret <- n + 2
                raise Return
            if mi = 2 then
                __ret <- (2 * n) + 3
                raise Return
            if mi = 3 then
                let nb = bit_len n
                if nb > 64 then
                    err <- ("A(m,n) had n of " + (string nb)) + " bits; too large"
                    __ret <- 0 :?> bigint
                    raise Return
                let r = pow_big (2 :?> bigint) (int n)
                __ret <- (8 * r) - 3
                raise Return
        if (bit_len n) = 0 then
            __ret <- ackermann2 (m - (1 :?> bigint)) (1 :?> bigint)
            raise Return
        __ret <- ackermann2 (m - (1 :?> bigint)) (ackermann2 m (n - (1 :?> bigint)))
        raise Return
        __ret
    with
        | Return -> __ret
and show (m: int) (n: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable m = m
    let mutable n = n
    try
        err <- ""
        let res = ackermann2 (m :?> bigint) (n :?> bigint)
        if err <> "" then
            printfn "%s" ((((("A(" + (string m)) + ", ") + (string n)) + ") = Error: ") + err)
            __ret <- ()
            raise Return
        if (bit_len res) <= 256 then
            printfn "%s" ((((("A(" + (string m)) + ", ") + (string n)) + ") = ") + (string res))
        else
            let s: string = string res
            let pre: string = s.Substring(0, 20 - 0)
            let suf: string = s.Substring((String.length s) - 20, (String.length s) - ((String.length s) - 20))
            printfn "%s" ((((((((("A(" + (string m)) + ", ") + (string n)) + ") = ") + (string (String.length s))) + " digits starting/ending with: ") + pre) + "...") + suf)
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        show 0 0
        show 1 2
        show 2 4
        show 3 100
        show 3 1000000
        show 4 1
        show 4 2
        show 4 3
        __ret
    with
        | Return -> __ret
main()
